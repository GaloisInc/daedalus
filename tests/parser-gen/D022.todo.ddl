-- test fragment from Jpeg

def Marker x = Match [ 0xFF, x ]

-- For families of markers, such as APP or SOF
def SomeMarker (front : uint 4) =
  block
    $[0xFF]
    let tag   = UInt8
    let upper = (tag >> 4) as! uint 4
    upper == front is true
    tag as! uint 4

def BE16 = UInt8 # UInt8

def NonZero P =
  block
    $$ = P;
    $$ > 0 is true;

def Payload P =
  block
    let size = BE16 as uint 64
    size >= 2 is true
    let len = size - 2
    let here = GetStream
    let there = Take len here
    SetStream (Take len here)
    $$ = P
    END
    SetStream (Drop len here)


-- Start / End image
def SOI = Marker 0xD8 <| Fail "Missing Start-of-Image"
def EOI = Marker 0xD9 <| Fail "Missing End-of-Image"



-- Comment
def COM =
  block
    Marker 0xFE
    Payload (Many UInt8)

-- Application specific
def APP (x : uint 4) P =
  block
    Marker (0xE # x)
    Payload P


-- Start of frame, uninterpreted
def SomeSOF =
  block
    sof  = SomeMarker 0xC
    sof == 4 is false
    data = SOFPayload

def SOFPayload =
  Payload
  block
    samplePrecision        = UInt8
    numberOfLines          = BE16
    numberOfSamplesPerLine = BE16
    let comNumber          = UInt8 as uint 64
    components             = Many comNumber FrameComponent


def FrameComponent =
  block
    identifier          = UInt8
    let samplingFactors = UInt8
    hoizontalSampling   = (samplingFactors >> 4) as! uint 4
    verticalSampling    = samplingFactors        as! uint 4
    quantTableSel       = UInt8

def DRI =
  block
    Marker 0xDD
    Payload BE16


-- Start of Scan SOS
def SOS =
  block
    Marker 0xDA
    header = Payload SOSHeader
    data   = Many EntropyEncodedEntry

def SOSHeader =
  block
    let componentNum = UInt8
    components = Many (componentNum as uint 64) SOSComponent
    ss = UInt8
    se = UInt8
    let a = UInt8
    ah = a >> 4 as! uint 4
    al = a      as! uint 4

def SOSComponent =
  block
    id      = UInt8
    let table  = UInt8
    acTable = table        as! uint 4
    dcTable = (table >> 4) as! uint 4

def SomeRST =
  block
    $$ = SomeMarker 0xD
    ($$ < 8) is true

def EntropyEncodedEntry =
  --First
  --  restart = SomeRST
  --  bytes   = Many (1..) EntropyEncodedByte
    Many (1..) EntropyEncodedByte

def EntropyEncodedByte =
  First
    block $$ = $[0xFF]; $[0x00]
    $[!0xFF]


def Segment =
  First
    comment = COM
    dri     = DRI
    -- sof     = SomeSOF
    -- sos     = SOS
    -- app     = SomeAPP
    -- dqt     = DQT
    -- dht     = DHT
    -- rst     = SomeRST

def SomeJpeg = block SOI; $$ = Many Segment; EOI

def Main = block SomeJpeg; END
