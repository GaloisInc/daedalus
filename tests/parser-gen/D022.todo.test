D022.todo.ddl
--dump-gen