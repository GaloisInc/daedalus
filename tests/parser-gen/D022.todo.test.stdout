Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _COM callsite _Segment (127,15)-(127,17)
    _Payload__122 callsite _COM (44,5)-(44,23)
    BE16 callsite _Payload__122 (23,16)-(23,19)
Start: BE16 (14,20)-(14,24)
SlkCfg{ q:10, ctrl:[?, ?, ?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(43),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _COM callsite _Segment (127,15)-(127,17)
    _Payload__122 callsite _COM (44,5)-(44,23)
    Many
Start: _Payload__122 (29,10)-(29,10)
SlkCfg{ q:106, ctrl:[?, ?, ?, ?, ?, ?, ?, *], sem:[?, ?, ], inp:--- }
DTrans [
  ( [(7),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
  ( [(16),]
  , EndInput
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _COM callsite _Segment (127,15)-(127,17)
    _Payload__122 callsite _COM (44,5)-(44,23)
Start: _Payload__122 (28,5)-(28,23)
SlkCfg{ q:95, ctrl:[?, ?, ?, ?, ?, ?, *], sem:[?, ?, ?, ], inp:--- }
DTrans [
  ( [(15),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
  ( [(24),]
  , EndInput
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _COM callsite _Segment (127,15)-(127,17)
    _Payload__122 callsite _COM (44,5)-(44,23)
Start: _Payload__122 (31,5)-(31,23)
SlkCfg{ q:123, ctrl:[?, ?, ?, ?, ?, ?, *], sem:[?, ?, ], inp:--- }
DTrans [
  ( [(46),(46),(44),]
  , HeadInput ['\255']
  , DTrans [
      ( [(21),]
      , HeadInput ['\217']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\221']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\254']
      , Resolution (NotAmbiguous)
      ),
    ]
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _COM callsite _Segment (127,15)-(127,17)
Start: _COM (44,5)-(44,23)
SlkCfg{ q:152, ctrl:[?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(17),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
    _Payload__121 callsite _DRI (81,5)-(81,16)
    BE16 callsite _Payload__121 (23,16)-(23,19)
Start: BE16 (14,20)-(14,24)
SlkCfg{ q:10, ctrl:[?, ?, ?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(43),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
    _Payload__121 callsite _DRI (81,5)-(81,16)
    _BE16 callsite _Payload__121 (29,10)-(29,10)
Start: _BE16 (14,20)-(14,24)
SlkCfg{ q:166, ctrl:[?, ?, ?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(10),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
    _Payload__121 callsite _DRI (81,5)-(81,16)
Start: _Payload__121 (28,5)-(28,23)
SlkCfg{ q:209, ctrl:[?, ?, ?, ?, ?, ?, *], sem:[?, ?, ?, ], inp:--- }
DTrans [
  ( [(15),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
    _Payload__121 callsite _DRI (81,5)-(81,16)
Start: _Payload__121 (30,5)-(30,7)
SlkCfg{ q:218, ctrl:[?, ?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(9),]
  , EndInput
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
    _Payload__121 callsite _DRI (81,5)-(81,16)
Start: _Payload__121 (31,5)-(31,23)
SlkCfg{ q:227, ctrl:[?, ?, ?, ?, ?, ?, *], sem:[?, ?, ], inp:--- }
DTrans [
  ( [(46),(46),(44),]
  , HeadInput ['\255']
  , DTrans [
      ( [(21),]
      , HeadInput ['\217']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\221']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\254']
      , Resolution (NotAmbiguous)
      ),
    ]
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
    _Segment callsite _SomeJpeg (136,37)-(136,43)
    _DRI callsite _Segment (128,15)-(128,17)
Start: _DRI (81,5)-(81,16)
SlkCfg{ q:256, ctrl:[?, ?, ?, ?, ?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(17),]
  , HeadInput ['\NUL','\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    Many
Start: _SomeJpeg (136,32)-(136,43)
SlkCfg{ q:294, ctrl:[?, ?, ?, *], sem:[?, ?, ], inp:--- }
DTrans [
  ( [(18),(18),(16),]
  , HeadInput ['\255']
  , DTrans [
      ( [(21),]
      , HeadInput ['\217']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\221']
      , Resolution (NotAmbiguous)
      ),
      ( [(9),]
      , HeadInput ['\254']
      , Resolution (NotAmbiguous)
      ),
    ]
  ),
]

Stack:
    *
    Main callsite __START__
    _SomeJpeg callsite Main (138,18)-(138,25)
    _SOI callsite _SomeJpeg (136,22)-(136,24)
    _Marker callsite _SOI (35,11)-(35,21)
Start: _Marker (3,16)-(3,32)
SlkCfg{ q:24, ctrl:[?, ?, ?, ?, *], sem:[], inp:--- }
DTrans [
  ( [(17),]
  , HeadInput ['\216']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
    Main callsite __START__
Start: Main (138,28)-(138,30)
SlkCfg{ q:318, ctrl:[?, *], sem:[?, ], inp:--- }
DTrans [
  ( [(9),]
  , EndInput
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
Start: Main (138,1)-(138,31)
SlkCfg{ q:1, ctrl:[*], sem:[*], inp:--- }
DTrans [
  ( [(20),]
  , HeadInput ['\255']
  , Resolution (NotAmbiguous)
  ),
]

Stack:
    *
Start: __START__
SlkCfg{ q:0, ctrl:[*], sem:[?, *], inp:--- }
DTrans [
  ( [(0),]
  , AcceptingPath
  , NotAmbiguous
  ),
]

**********************
***** Strict LLA *****
**********************
("DataDepTrans",4)
("Lookahead-0",1)
("Lookahead-1",12)
("Lookahead-2",3)

Total nb DFAs: 20

SUCCESS: Fully determinized format
